# Task 1 — Data Cleaning & Preprocessing

## Netflix Dataset

### Objective

The objective of this task is to clean, transform, and prepare the Netflix dataset for further analysis.

### Workflow

1. Import the dataset using Pandas.
2. Identify missing and duplicate records.
3. Handle null values and inconsistent formats.
4. Transform categorical and date-related columns.
5. Create a clean dataset for further analysis.

### Tools & Libraries

- Python
- Pandas
- Jupyter Notebook

In [1]:
import pandas as pd
import numpy as np

## 1. Import the Dataset

The Netflix dataset is loaded using Pandas. We first inspect its structure, dimensions, and initial records to understand the raw data.

In [2]:
df = pd.read_csv("../data/Dataset.csv")

df.head()

,show_id,type,title,director,country,date_added,release_year,rating,duration,listed_in
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,United States,9/25/2021,2020,PG-13,90 min,Documentaries
1,s3,TV Show,Ganglands,Julien Leclercq,France,9/24/2021,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act..."
2,s6,TV Show,Midnight Mass,Mike Flanagan,United States,9/24/2021,2021,TV-MA,1 Season,"TV Dramas, TV Horror, TV Mysteries"
3,s14,Movie,Confessions of an Invisible Girl,Bruno Garotti,Brazil,9/22/2021,2021,TV-PG,91 min,"Children & Family Movies, Comedies"
4,s8,Movie,Sankofa,Haile Gerima,United States,9/24/2021,1993,TV-MA,125 min,"Dramas, Independent Movies, International Movies"


In [3]:
print(f"Dataset contains {df.shape[0]} rows and {df.shape[1]} columns.")

Dataset contains 8790 rows and 10 columns.


### Dataset Structure

Let's examine the column names and data types before performing any cleaning operations.

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8790 entries, 0 to 8789
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8790 non-null   object
 1   type          8790 non-null   object
 2   title         8790 non-null   object
 3   director      8790 non-null   object
 4   country       8790 non-null   object
 5   date_added    8790 non-null   object
 6   release_year  8790 non-null   int64 
 7   rating        8790 non-null   object
 8   duration      8790 non-null   object
 9   listed_in     8790 non-null   object
dtypes: int64(1), object(9)
memory usage: 686.8+ KB


In [5]:
df.columns.tolist()

['show_id',
 'type',
 'title',
 'director',
 'country',
 'date_added',
 'release_year',
 'rating',
 'duration',
 'listed_in']

In [6]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
show_id,8790,8790,s1,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
type,8790,2,Movie,6126,NaN,NaN,NaN,NaN,NaN,NaN,NaN
title,8790,8787,9-Feb,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
director,8790,4528,Not Given,2588,NaN,NaN,NaN,NaN,NaN,NaN,NaN
country,8790,86,United States,3240,NaN,NaN,NaN,NaN,NaN,NaN,NaN
date_added,8790,1713,1/1/2020,110,NaN,NaN,NaN,NaN,NaN,NaN,NaN
release_year,8790.0,NaN,NaN,NaN,2014.183163,8.825466,1925.0,2013.0,2017.0,2019.0,2021.0
rating,8790,14,TV-MA,3205,NaN,NaN,NaN,NaN,NaN,NaN,NaN
duration,8790,220,1 Season,1791,NaN,NaN,NaN,NaN,NaN,NaN,NaN
listed_in,8790,513,"Dramas, International Movies",362,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## 2. Identify Missing Values

Missing values can affect data analysis and machine learning models. Therefore, we identify the number of missing values in each column before deciding how to handle them.

In [7]:
missing_values = df.isnull().sum()

missing_values

show_id         0
type            0
title           0
director        0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
dtype: int64

In [8]:
missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_summary = pd.DataFrame({
    "Missing Values": missing_values,
    "Missing Percentage": missing_percentage
})

missing_summary

,Missing Values,Missing Percentage
show_id,0,0.0
type,0,0.0
title,0,0.0
director,0,0.0
country,0,0.0
date_added,0,0.0
release_year,0,0.0
rating,0,0.0
duration,0,0.0
listed_in,0,0.0


### Duplicate Records

Duplicate records can lead to biased analysis and incorrect statistics. We check for duplicate rows before creating the cleaned dataset.

In [9]:
duplicate_count = df.duplicated().sum()

print(f"Number of duplicate rows: {duplicate_count}")

Number of duplicate rows: 0


In [10]:
df[df.duplicated()]

,show_id,type,title,director,country,date_added,release_year,rating,duration,listed_in


### Duplicate Treatment

No duplicate records were found in the dataset, so no duplicate rows were removed.

In [11]:
text_columns = [
    "type",
    "title",
    "director",
    "country",
    "rating",
    "duration",
    "listed_in"
]

In [12]:
for column in text_columns:
    df[column] = df[column].astype("string").str.strip()

In [13]:
df["director"].value_counts().head(10)

director
Not Given                 2588
Rajiv Chilaka               20
Alastair Fothergill         18
Raúl Campos, Jan Suter      18
Suhas Kadav                 16
Marcus Raboy                16
Jay Karas                   14
Cathy Garcia-Molina         13
Youssef Chahine             12
Jay Chapman                 12
Name: count, dtype: Int64

In [14]:
(df == "Not Given").sum()

show_id            0
type               0
title              0
director        2588
country          287
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
dtype: Int64

In [15]:
df = df.replace("Not Given", pd.NA)

In [16]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2588
country          287
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
dtype: int64

### Standardizing Missing Values

Some fields use the text value `"Not Given"` to represent unavailable information. These placeholder values are converted to proper missing values (`NaN`) to maintain consistency during analysis.

In [17]:
df["date_added"] = pd.to_datetime(
    df["date_added"],
    errors="coerce"
)

In [18]:
df["date_added"].dtype

dtype('<M8[ns]')

In [19]:
print("Invalid dates:", df["date_added"].isna().sum())

Invalid dates: 0


In [20]:
df["added_year"] = df["date_added"].dt.year
df["added_month"] = df["date_added"].dt.month

In [21]:
df[["date_added", "added_year", "added_month"]].head()

,date_added,added_year,added_month
0,2021-09-25,2021,9
1,2021-09-24,2021,9
2,2021-09-24,2021,9
3,2021-09-22,2021,9
4,2021-09-24,2021,9


In [22]:
df["duration"].value_counts().head(20)

duration
1 Season     1791
2 Seasons     421
3 Seasons     198
90 min        152
97 min        146
93 min        146
94 min        146
91 min        144
95 min        137
96 min        130
92 min        129
102 min       122
98 min        120
99 min        118
101 min       116
88 min        116
103 min       114
106 min       111
100 min       108
89 min        106
Name: count, dtype: Int64

In [23]:
df["duration_value"] = (
    df["duration"]
    .str.extract(r"(\d+)")
    .astype("Int64")
)

In [24]:
df[["type", "duration", "duration_value"]].head(10)

,type,duration,duration_value
0,Movie,90 min,90
1,TV Show,1 Season,1
2,TV Show,1 Season,1
3,Movie,91 min,91
4,Movie,125 min,125
5,TV Show,9 Seasons,9
6,Movie,104 min,104
7,Movie,87 min,87
8,Movie,127 min,127
9,Movie,76 min,76


In [25]:
df["duration_unit"] = df["duration"].str.extract(r"([A-Za-z]+)")

In [26]:
df[["duration", "duration_value", "duration_unit"]].head(10)

,duration,duration_value,duration_unit
0,90 min,90,min
1,1 Season,1,Season
2,1 Season,1,Season
3,91 min,91,min
4,125 min,125,min
5,9 Seasons,9,Seasons
6,104 min,104,min
7,87 min,87,min
8,127 min,127,min
9,76 min,76,min


In [27]:
categorical_columns = [
    "type",
    "rating",
    "duration_unit"
]

In [28]:
for column in categorical_columns:
    print(f"\n{column}:")
    print(df[column].value_counts().head(10))


type:
type
Movie      6126
TV Show    2664
Name: count, dtype: Int64

rating:
rating
TV-MA    3205
TV-14    2157
TV-PG     861
R         799
PG-13     490
TV-Y7     333
TV-Y      306
PG        287
TV-G      220
NR         79
Name: count, dtype: Int64

duration_unit:
duration_unit
min        6126
Season     1791
Seasons     873
Name: count, dtype: Int64


In [29]:
print("Final Shape:", df.shape)

Final Shape: (8790, 14)


In [30]:
df.isnull().sum()

show_id              0
type                 0
title                0
director          2588
country            287
date_added           0
release_year         0
rating               0
duration             0
listed_in            0
added_year           0
added_month          0
duration_value       0
duration_unit        0
dtype: int64

In [31]:
print("Duplicates:", df.duplicated().sum())

Duplicates: 0


In [32]:
df.dtypes

show_id                   object
type              string[python]
title             string[python]
director          string[python]
country           string[python]
date_added        datetime64[ns]
release_year               int64
rating            string[python]
duration          string[python]
listed_in         string[python]
added_year                 int32
added_month                int32
duration_value             Int64
duration_unit     string[python]
dtype: object

In [33]:
df.head()

,show_id,type,title,director,country,date_added,release_year,rating,duration,listed_in,added_year,added_month,duration_value,duration_unit
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,United States,2021-09-25,2020,PG-13,90 min,Documentaries,2021,9,90,min
1,s3,TV Show,Ganglands,Julien Leclercq,France,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",2021,9,1,Season
2,s6,TV Show,Midnight Mass,Mike Flanagan,United States,2021-09-24,2021,TV-MA,1 Season,"TV Dramas, TV Horror, TV Mysteries",2021,9,1,Season
3,s14,Movie,Confessions of an Invisible Girl,Bruno Garotti,Brazil,2021-09-22,2021,TV-PG,91 min,"Children & Family Movies, Comedies",2021,9,91,min
4,s8,Movie,Sankofa,Haile Gerima,United States,2021-09-24,1993,TV-MA,125 min,"Dramas, Independent Movies, International Movies",2021,9,125,min


In [34]:
output_path = "../data/netflix_cleaned.csv"

df.to_csv(output_path, index=False)

print(f"Cleaned dataset saved to: {output_path}")

Cleaned dataset saved to: ../data/netflix_cleaned.csv


## 4. Final Summary

The Netflix dataset was successfully cleaned and prepared for further analysis.

### Data Cleaning Results

- Dataset imported successfully using Pandas.
- Missing values were identified and standardized.
- Duplicate records were checked.
- `"Not Given"` placeholder values were standardized as missing values.
- Text columns were cleaned by removing unnecessary whitespace.
- `date_added` was converted to a datetime format.
- Year and month features were extracted from `date_added`.
- The mixed `duration` column was transformed into numeric duration and duration unit features.
- The cleaned dataset was exported as `netflix_cleaned.csv`.

The resulting dataset is now ready for exploratory data analysis and further data science tasks.

In [35]:
print("Final Shape:", df.shape)
print("Duplicate Records:", df.duplicated().sum())

print("\nMissing Values:")
print(df.isnull().sum())

print("\nData Types:")
print(df.dtypes)

print("\nFirst 5 Rows:")
display(df.head())

Final Shape: (8790, 14)
Duplicate Records: 0

Missing Values:
show_id              0
type                 0
title                0
director          2588
country            287
date_added           0
release_year         0
rating               0
duration             0
listed_in            0
added_year           0
added_month          0
duration_value       0
duration_unit        0
dtype: int64

Data Types:
show_id                   object
type              string[python]
title             string[python]
director          string[python]
country           string[python]
date_added        datetime64[ns]
release_year               int64
rating            string[python]
duration          string[python]
listed_in         string[python]
added_year                 int32
added_month                int32
duration_value             Int64
duration_unit     string[python]
dtype: object

First 5 Rows:


,show_id,type,title,director,country,date_added,release_year,rating,duration,listed_in,added_year,added_month,duration_value,duration_unit
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,United States,2021-09-25,2020,PG-13,90 min,Documentaries,2021,9,90,min
1,s3,TV Show,Ganglands,Julien Leclercq,France,2021-09-24,2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",2021,9,1,Season
2,s6,TV Show,Midnight Mass,Mike Flanagan,United States,2021-09-24,2021,TV-MA,1 Season,"TV Dramas, TV Horror, TV Mysteries",2021,9,1,Season
3,s14,Movie,Confessions of an Invisible Girl,Bruno Garotti,Brazil,2021-09-22,2021,TV-PG,91 min,"Children & Family Movies, Comedies",2021,9,91,min
4,s8,Movie,Sankofa,Haile Gerima,United States,2021-09-24,1993,TV-MA,125 min,"Dramas, Independent Movies, International Movies",2021,9,125,min


## Task 1 Completion

The Netflix dataset has been successfully cleaned and preprocessed.

The final dataset contains 8,790 records and 14 features. Duplicate records were removed/verified, missing values were identified and standardized, text fields were cleaned, date information was converted into a proper datetime format, and additional date and duration features were created.

The cleaned dataset is now ready for Exploratory Data Analysis (EDA) and further data science tasks.